### ISSUE TITLE
Shell script improvements (#771)

### ISSUE THREAD

**Issue by aarondill** [pc2ccs/pc2v9#775]
**Describe the issue**: please provide a clear and concise description of what the bug is.  

This is an issue to correspond with PR #771, which makes the shell scripts better handle spaces in file names.

**To Reproduce**: please describe, in order, the steps to reproduce the behavior:

```console
> git clone 'https://github.com/pc2ccs/pc2v9'
> cd pc2v9 
> mv bin 'binary files' # this is the cause of the bugs
> ./bin/pc2team # or any script that calls pc2env
./binary files/pc2team: 9: .: .: not found
```

**Expected behavior**: please provide a description of what you *expected* to happen when you took the above steps:  

This program should run as normal.

**Actual behavior**: please describe what _actually_ happened when you took the above steps:

The program fails because `dirname $0` evaluates to `dirname './binary' 'files/pc2team`, which in turn is `. files`, which then becomes `. . files/pc2env`, where `files/pc2env` is *passed as an argument* to the nonexistent `.` script and thus, the program fails.

**Environment**: please complete the following information:
 - OS name and version under which the issue was detected (e.g. Windows 10 or Ubuntu 18.04): Ubuntu 22.04.2 LTS
 - PC2 Version (e.g. 9.7.0):   9.8.0 (and/or `master`)
 - Browser (if the bug is related to web-based team clients) (e.g. _Chrome Version 75.0.3770.142_ or _Safari Version 11.1.2_): N/A

**Log Info**:  PC^2 generates logs (in the **logs** folder for client modules; under the **profiles** folder for the PC2 Server).  Please review the relevant logs and provide copies of them as attachments -- particularly if the logs show Exceptions being thrown.  

This issue occurs before the pc2 software is started, and thus no logs are generated.

**Additional context**: please add any other context about the issue here.  

It seems my statement about the legacy ``` `backtick` ``` syntax was slightly misunderstood. It is not necessarily *safer* than the `$(command substitution)` syntax, however, it has been depreciated, and is strongly encouraged to move away from the former. Additionally, it has several issues, least of which is readability as ``` `cmd 'arg' "$var" \`other_cmd "arg$var"\`` ``` can quickly become confusing, but additionally has a few other inconsistencies that are resolved with the `$()` notation. [Here](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_xcu_chap02.html#tag_23_02_06_03) is a article from the Open Group Base, and [here](https://www.redhat.com/sysadmin/backtick-operator-vs-parens) is a slightly easier to read article from Red Hat about this issue.

**Comment by aarondill** [comment]
@lane55 Here is the corresponding issue for #771

**Comment by lane55** [comment]
The test case submitted, would nearly never happen.
> cd pc2v9 
> mv bin 'binary files' # this is the cause of the bugs
> ./bin/pc2team # or any script that calls pc2env
./binary files/pc2team: 9: .: .: not found
You may notice that we use relative paths in the scripts, pc2env, this handles most cases of spaces in parent dirctories.

**Comment by lane55** [comment]
I agree the backtick changes are be a good security fix.

**Comment by lane55** [comment]
With reguard to changs ot the command line, can you provide a test case/exmple where the PR changes would be needed ?

**Comment by aarondill** [comment]
Here's a minimal example, unrelated specifically to the pc2 project, but important for demonstration (same issue applies with the pc2 scripts)

```console
> mkdir 'folder with spaces'
> cd 'folder with spaces'
> echo $'#!/bin/sh\nprintf "output: %s\\n" $0 # unquoted!' > executable
> cat executable 
#!/bin/sh
printf "output: %s" $0 # unquoted!
> chmod +x executable
> ./executable
output: ./executable
> "$PWD/executable"
output: /home/aaron/folder
output: with
output: spaces/executable
```

**Comment by aarondill** [comment]
This is a risk anytime `$0` is used in a script, and is one of the major reasons *quotes are very necessary* in any shell script with user input (including any positional argument).

**Comment by lane55** [comment]
I agree it is a security risk, the risk is greater than zero.   Then ther are a number of other considerations.
1 - is the data sensitive personal information? (no)
2 - once the pc2 server starts, is there a risk from these potential exploits? (no)
3 - is pc2 a high value target?  (no)
4 - what is the risk if such an exploit were done?   That would require a lot of knowledge of how pc2 works
  to use that exploit to make any changes.
So the changs are a good idea, but there are not a critial security risk.

**Comment by lane55** [comment]
> Here's a minimal example, unrelated specifically to the pc2 project, but important for demonstration (same issue applies with the pc2 scripts)
> 
> ```
> > mkdir 'folder with spaces'
> > cd 'folder with spaces'
> > echo $'#!/bin/sh\nprintf "output: %s\\n" $0 # unquoted!' > executable
> > cat executable 
> #!/bin/sh
> printf "output: %s" $0 # unquoted!
> > chmod +x executable
> > ./executable
> output: ./executable
> > "$PWD/executable"
> output: /home/aaron/folder
> output: with
> output: spaces/executable
> ```

Each of our team numbers undestands the exploits, each team member have more than 20 years developing software,
so we understand the suggested fix as well as risk.

**Comment by lane55** [comment]
@aarondill would you like to talk about these changes?   There are  additional comments we/I  have and if you want
to learn about them and move this PR through the process, it would be good to talk.

**Comment by aarondill** [comment]
> @aarondill would you like to talk about these changes?   There are  additional comments we/I  have and if you want
> 
> to learn about them and move this PR through the process, it would be good to talk.

Of course, I'd be more than happy to continue discussions about these changes. If there's an additional step in the process that i've missed while reading the contribution guide, please let me know and i'd be happy to remedy the issue!

It's also good to point out that while most of these changes *could potentially* improve security in one way or another, as any exploits using them would require access directly to the shell and would be daily pointless anyways, that these are really just edge-case avoidance changes, as well as quality-of-life improvements for UNIX users who may (for example) wish to place these scripts in their PATH (or an alias) to allow running from anywhere.
